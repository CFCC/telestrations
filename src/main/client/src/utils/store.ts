import { createGamekitSlice, sendEvent } from "@piticent123/gamekit-client";
import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import { GameState, TelestrationsGame, TelestrationsPlayer } from "./types";
import _ from "lodash";
import { Settings } from "@piticent123/gamekit-client/lib/types";

const { reducer: gamekitReducer, actions: gamekitActions } =
  createGamekitSlice<Settings, TelestrationsPlayer, TelestrationsGame>();

interface AppState {
  gameState: GameState;
  activeContentId: string;
}
interface State {
  app: AppState;
  gamekit: ReturnType<typeof gamekitReducer>;
}
interface ThunkApi {
  state: State;
}

// This has to be at the top level to be throttled correctly
const _sendWritePageEvent = _.throttle((gameCode: string, content: string) => {
  sendEvent({
    route: `/app/games/${gameCode}/page/write`,
    data: content,
  });
}, 500);
export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    _sendWritePageEvent(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({
      route: `/app/games/${currentGame.code}/page/submit`,
      data: content,
    });
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (__, { getState }) => {
    const {
      gamekit: { currentGame },
    } = getState();

    sendEvent({ route: `/games/${currentGame.code}/start` });
  }
);

export const {
  createGame,
  joinGame,
  rejoinGame,
  connectToServer,
  saveSettings,
  becomeAdmin,
} = gamekitActions;

const { actions, reducer: appReducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    activeContentId: "",
  } as AppState,
  reducers: {
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(gamekitActions.rejoinGame.fulfilled, (state) => {
        state.gameState = GameState.IN_GAME;
      })
      .addCase(gamekitActions.connectToServer.fulfilled, (state) => {
        state.gameState = GameState.GAME_CODE;
      })
      .addCase(gamekitActions.joinGame.fulfilled, (state) => {
        state.gameState = GameState.WAITING_TO_START;
      })
      .addCase(submitGuess.fulfilled, (state) => {
        state.gameState = GameState.WAITING_FOR_CONTENT;
      });
  },
});

export const {
  viewNotepadHistory,
  viewPlayerHistory,
  setGameState
} = actions;

export const store = configureStore({
  reducer: { gamekit: gamekitReducer, app: appReducer },
});

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
