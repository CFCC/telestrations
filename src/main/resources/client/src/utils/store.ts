import {
  configureStore,
  createAsyncThunk,
  createSlice,
  PayloadAction,
} from "@reduxjs/toolkit";
import {
  TypedUseSelectorHook,
  useSelector as useUntypedSelector,
} from "react-redux";

import * as api from "./api";
import { Game, GameState, Settings } from "./types";

interface State {
  gameState: GameState;
  openGames: string[];
  toast: {
    id: number;
    title: string;
    description: string;
    status: "info" | "warning" | "success" | "error";
  };
  settings: Settings;
  currentGame: Game;
  activeContentId: string;
}
interface ThunkApi {
  state: State;
}

export const goToLobby = createAsyncThunk<void, void, ThunkApi>(
  "goToLobby",
  (_, { getState }) => {
    const { settings } = getState();
    api.connectToServer(settings.id);
  }
);

export const createAndJoinGame = createAsyncThunk<void, string, ThunkApi>(
  "createGame",
  async (code, { getState }) => {
    const { settings } = getState();
    await api.createGame(code);
    await api.joinGame(code, settings);
  }
);

export const joinGame = createAsyncThunk<void, string, ThunkApi>(
  "joinGame",
  async (code, { getState }) => {
    const { settings } = getState();
    await api.joinGame(code, settings);
  }
);

export const rejoinGame = createAsyncThunk<void, string, ThunkApi>(
  "rejoinGame",
  async (code, { getState }) => {
    if (!code) return;

    const { settings } = getState();
    await api.joinGame(code, settings, true);
  }
);

export const saveSettings = createAsyncThunk<
  Partial<Settings>,
  Partial<Settings>,
  ThunkApi
>("saveSettings", (settings, { getState }) => {
  const {
    currentGame: { code },
    settings: oldSettings,
  } = getState();

  api.updateSettings(code, { ...oldSettings, ...settings });
  Object.entries(settings).forEach(([k, v]) => {
    if (!v) return;
    localStorage.setItem(k, v.toString());
  });

  return settings;
});

export const setGuess = createAsyncThunk<void, string, ThunkApi>(
  "setGuess",
  async (content, { getState }) => {
    const { currentGame } = getState();
    await api.setGuess(currentGame.code, content);
  }
);

export const submitGuess = createAsyncThunk<void, string, ThunkApi>(
  "submitGuess",
  async (content, { getState }) => {
    const { currentGame } = getState();
    await api.submitGuess(currentGame.code, content);
  }
);

export const startGame = createAsyncThunk<void, string, ThunkApi>(
  "startGame",
  async (gameCode) => {}
);

export const { actions, reducer } = createSlice({
  name: "app",
  initialState: {
    gameState: GameState.LOGIN,
    openGames: [],
    toast: { id: 0, title: "", description: "", status: "info" },
    currentGame: {
      active: false,
      code: "",
      players: [],
      admin: "",
      round: 0,
      adminId: "",
      isDone: false,
    },
    settings: {
      id: localStorage.getItem("id"),
      avatar: localStorage.getItem("avatar"),
      name: localStorage.getItem("name"),
      connected: true,
    },
    activeContentId: "",
  } as State,
  reducers: {
    handleRequestException: (state, action: PayloadAction<string>) => {
      state.toast = {
        id: state.toast.id + 1,
        title: "Bad Request",
        description: action.payload,
        status: "error",
      };
    },
    handleServerException: (state, action: PayloadAction<string>) => {
      state.toast = {
        id: state.toast.id + 1,
        title: "Server Error",
        description: action.payload,
        status: "error",
      };
    },
    handleSuccess: (state, action: PayloadAction<string>) => {
      state.toast = {
        id: state.toast.id + 1,
        title: "",
        description: action.payload,
        status: "success",
      };
    },
    handleGamesListMessage: (state, action: PayloadAction<string[]>) => {
      state.openGames = [...action.payload];
    },
    handleGameUpdate: (state, action: PayloadAction<Game>) => {
      state.currentGame = action.payload;
    },
    viewPlayerHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.PLAYER_HISTORY;
      state.activeContentId = action.payload;
    },
    viewNotepadHistory: (state, action: PayloadAction<string>) => {
      state.gameState = GameState.NOTEPAD_HISTORY;
      state.activeContentId = action.payload;
    },
    setGameState: (state, action: PayloadAction<GameState>) => {
      state.gameState = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(saveSettings.fulfilled, (state, action) => {
        state.settings = {
          ...state.settings,
          ...action.payload,
        };
      })
      .addCase(rejoinGame.fulfilled, (state, action) => {
        state.gameState = GameState.IN_GAME;
      });
  },
});

export const store = configureStore({ reducer });

export const useSelector: TypedUseSelectorHook<State> = useUntypedSelector;
