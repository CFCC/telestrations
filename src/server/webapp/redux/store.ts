import {createStore} from '';
import reducer from '../reducers';

const store = createStore(reducer,
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__());

export default store;