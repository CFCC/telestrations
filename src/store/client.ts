import {createContext} from "react";
import firebase from "firebase";

import {Action, Actions, ActionTypes, NewContentDTO, State, Store} from "./client.types";
import {ClientGameState} from "../types/client";
import {ContentType} from "../types/shared";
import {finishTurn, joinGame, updateGuess} from "../firebase-client/client";

export const initialState: State = {
    user: null,
    gameCode: "",
    gameState: ClientGameState.LOGIN,
    guess: "",
    content: "",
};

export function reducer(state: State, action: Action): State {
    switch (action.type) {
        case ActionTypes.SET_USER:
            return {...state, user: action.user, gameState: ClientGameState.GAME_SELECTION};
        case ActionTypes.JOIN_GAME:
            joinGame(state.user, action.gameCode);
            return {...state, gameState: ClientGameState.WAITING_TO_START};
        case ActionTypes.SET_GUESS:
            updateGuess(state.user, state.notepadId, state.gameCode, action.guess);
            return {...state, guess: action.guess};
        case ActionTypes.SUBMIT_GUESS:
            finishTurn(state.user, state.gameCode);
            return state;
        case ActionTypes.NEW_CONTENT:
            return {
                ...state,
                gameState: action.content.type === ContentType.Text ? ClientGameState.DRAWING : ClientGameState.TYPING,
                content: action.content.content,
                guess: "",
            };
        default:
            return state;
    }
}

export const actionCreators: Actions = {
    setUser: (user: firebase.User | null) => ({type: ActionTypes.SET_USER, user}),
    newContent: (content: NewContentDTO) => ({type: ActionTypes.NEW_CONTENT, content}),
    joinGame: (gameCode: string) => ({type: ActionTypes.JOIN_GAME, gameCode}),
    setGuess: (guess: string) => ({type: ActionTypes.SET_GUESS, guess}),
    submitGuess: () => ({type: ActionTypes.SUBMIT_GUESS}),
};

export const GameContext = createContext([initialState, actionCreators] as Store);
