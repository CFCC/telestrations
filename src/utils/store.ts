import {configureStore, createSlice, PayloadAction} from "@reduxjs/toolkit";
import {TypedUseSelectorHook, useSelector as useUntypedSelector} from "react-redux";

import {User, Game, Notepad, Player, WithId} from "./firebase";

export enum GameState {
    // Common
    LOGIN = "login",
    GAME_CODE = "game code",
    WAITING_TO_START = "waiting to start",

    // Server
    BIRDS_EYE = "bird's eye",
    SINGLE_PLAYER = "single player",
    PLAYER_HISTORY = "player history",
    NOTEPAD_HISTORY = "notepad history",

    // Client
    IN_GAME = "in game",
    FINISHED = "finished",
    WAITING_FOR_CONTENT = "waiting for content",
}

export const firebaseSlice = createSlice({
    name: "firebase",
    initialState: {
        game: {
            id: "",
            created: new Date().getTime(),
            status: "lobby",
            serverId: "",
        },
        players: {} as Record<string, Player>,
        notepads: {} as Record<string, Notepad>,
    },
    reducers: {
        updateGame: (state, action: PayloadAction<Partial<WithId<Game>>>) => {
            state.game = {...state.game, ...action.payload};
        },
        updateNotepads: (state, action: PayloadAction<Record<string, Notepad>>) => {
            state.notepads = action.payload;
        },
        updatePlayers: (state, action: PayloadAction<Record<string, Player>>) => {
            state.players = action.payload;
        }
    }
});
export const clientSlice = createSlice({
    name: "client",
    initialState: {
        user: {displayName: "", uid: ""} as User,
        gameState: GameState.LOGIN,
        activePlayerId: "",
    },
    reducers: {
        viewPlayerHistory: (state, action: PayloadAction<string>) => {
            state.gameState = GameState.PLAYER_HISTORY;
            state.activePlayerId = action.payload;
        },
        viewNotepadHistory: (state, action: PayloadAction<string>) => {
            state.gameState = GameState.NOTEPAD_HISTORY;
            state.activePlayerId = action.payload;
        },
        setUser: (state, action: PayloadAction<User>) => {
            state.user = action.payload;
            state.gameState = GameState.GAME_CODE;
        },
        setGameState: (state, action: PayloadAction<GameState>) => {
            state.gameState = action.payload;
        }
    },
    extraReducers: builder => builder
});

export const store = configureStore({
    reducer: {
        firebase: firebaseSlice.reducer,
        client: clientSlice.reducer
    }
});

export const useSelector: TypedUseSelectorHook<ReturnType<typeof store.getState>> = useUntypedSelector;
