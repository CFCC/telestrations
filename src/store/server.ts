import _ from "lodash";
import { createContext } from "react";

import * as firebase from "firebase-client/server";
import { ServerGameState, ServerPlayer, PlayerDTO, NotepadPageDTO } from "types/server";
import { Notepad } from "types/client";
import { UUID } from "types/shared";
import { State, Action, ActionTypes, viewPlayerHistory, viewNotepadHistory, startGame, addPlayer, updateGuess, finishedGameTurn, gameFinished, Store, Actions, setGameCode } from "store/server.types";

export const initialState: State = {
    gameState: ServerGameState.GAME_CODE,
    serverId: "",
    gameCode: "",
    players: [],
    notepads: [],
    activePlayerId: "",
    activeNotepadId: "",
};

const defaultPlayer: ServerPlayer = {
    id: "",
    nickname: "",
    queueOfOwners: [],
    notepadIndex: 0,
    ownerOfCurrentNotepad: "",
};

const defaultNotepad: Notepad = {
    owner: "",
    content: [],
};

export function reducer(state: State, action: Action): State {
    switch (action.type) {
        case ActionTypes.SET_GAME_CODE:
            firebase.addGameToLobby(action.gameCode);
            return {...state, gameCode: action.gameCode, gameState: ServerGameState.LOADING};
        case ActionTypes.SET_SERVER_ID:
            return {...state, serverId: action.serverId};
        case ActionTypes.START_GAME:
            // io.startGame();
            return {
                ...state,
                gameState: ServerGameState.BIRDS_EYE,
                notepads: state.players.map(p => ({owner: p.id, content: []})),
            };
        case ActionTypes.PLAYER_ADDED: {
            const players = state.players.slice(0);
            players.push({ownerOfCurrentNotepad: action.player.id, notepadIndex: 0, queueOfOwners: [], ...action.player});
            return {...state, players};
        }
        case ActionTypes.UPDATE_GUESS: {
            const notepads = state.notepads.slice(0);
            const player = state.players.find(p => p.id === action.playerId) || defaultPlayer;
            const notepad = notepads.find(n => n.owner === player.ownerOfCurrentNotepad) || defaultNotepad;
            notepad.content[player.notepadIndex] = action.content;
            return {...state, notepads};
        }
        case ActionTypes.FINISHED_GAME_TURN: {
            const players = state.players.slice(0);
            Object.assign(players[_.findIndex(players, {id: action.playerId})], {
                ownerOfCurrentNotepad: "",
                notepadIndex: -1,
            });
            return {...state, players};
        }
        case ActionTypes.NEW_NOTEPAD: {
            const players = state.players.slice(0);
            const newNotepad = _.find(state.notepads, {owner: action.newNotepadOwnerId});
            Object.assign(players[_.findIndex(players, {id: action.playerId})], {
                ownerOfCurrentNotepad: action.newNotepadOwnerId,
                notepadIndex: newNotepad ? newNotepad.content.length : -1,
            });
            return {...state, players};
        }
        default:
            return state;
    }
}

export const actionCreators: Actions = {
    setGameCode: (gameCode: string) => ({type: ActionTypes.SET_GAME_CODE, gameCode} as setGameCode),
    viewPlayerHistory: (playerId: UUID) => ({type: ActionTypes.VIEW_PLAYER_HISTORY, playerId} as viewPlayerHistory),
    viewNotepadHistory: (ownerId: UUID) => ({type: ActionTypes.VIEW_NOTEPAD_HISTORY, ownerId} as viewNotepadHistory),
    startGame: () => ({type: ActionTypes.START_GAME} as startGame),
    addPlayer: (player: PlayerDTO) => ({type: ActionTypes.PLAYER_ADDED, player} as addPlayer),
    updateGuess: ({playerId, content}: NotepadPageDTO) => ({type: ActionTypes.UPDATE_GUESS, playerId, content} as updateGuess),
    finishedGameTurn: (playerId: UUID) => ({type: ActionTypes.FINISHED_GAME_TURN, playerId} as finishedGameTurn),
    gameFinished: () => ({type: ActionTypes.GAME_FINISHED} as gameFinished),
    // newNotepad: ({playerId, newNotepadOwnerId}: FinishedGameTurnDTO) => ({type: ActionTypes.NEW_NOTEPAD, playerId, newNotepadOwnerId} as newNotepad),
};

export const GameContext = createContext([initialState, actionCreators] as Store);