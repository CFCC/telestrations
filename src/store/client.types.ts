import { ReactNode } from "react";

import { ClientGameState } from "types/client";
import { NewContentDTO } from "types/server";

export interface State extends Record<string, any> {
    user: firebase.User | null;
    gameCode: string;
    gameState: ClientGameState;
    guess: string;
    content: string;
}

export enum ActionTypes {
    SET_USER = "SET_USER",
    NEW_CONTENT = "NEW_CONTENT",
    SET_GAME_STATE = "SET_GAME_STATE",
    SUBMIT_NICKNAME = "SUBMIT_NICKNAME",
    SET_GUESS = "SET_GUESS",
    SUBMIT_GUESS = "SUBMIT_GUESS",
    INIT = "INIT",
}

export interface setUser {
    type: ActionTypes.SET_USER;
    user: firebase.User | null;
}

export interface newContent {
    type: ActionTypes.NEW_CONTENT;
    content: NewContentDTO
}

export interface setGameState {
    type: ActionTypes.SET_GAME_STATE;
    state: ClientGameState;
}

export interface submitNickname {
    type: ActionTypes.SUBMIT_NICKNAME;
    nickname: String;
}

export interface setGuess {
    type: ActionTypes.SET_GUESS;
    guess: string;
}

export interface submitGuess {
    type: ActionTypes.SUBMIT_GUESS;
}

export interface init {
    type: ActionTypes.INIT;
}

export type Action = setUser | setGameState | submitNickname | setGuess | submitGuess | newContent | init;

export interface Actions extends Record<string, (...args: any) => Action> {
    setUser: (user: firebase.User | null) => setUser,
    newContent: (content: NewContentDTO) => newContent,
    setGameState: (state: ClientGameState) => setGameState,
    submitNickname: (nickname: String) => submitNickname,
    setGuess: (guess: string) => setGuess,
    submitGuess: () => submitGuess,
    init: () => init,
}

export type Store = [State, Actions];
