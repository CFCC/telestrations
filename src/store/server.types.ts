import {UUID} from "../types/shared";
import {ServerGameState} from "../types/server";
import {Game, Notepad, Player} from "../types/firebase";

export interface State {
    gameState: ServerGameState;
    game: Game;
    gameCode: string;
    activePlayerId: UUID;
    activeNotepadId: UUID;
}

export enum ActionTypes {
    SET_GAME_CODE = "SET_GAME_CODE",
    VIEW_PLAYER_HISTORY = "VIEW_PLAYER_HISTORY",
    VIEW_NOTEPAD_HISTORY = "VIEW_NOTEPAD_HISTORY",
    START_GAME = "START_GAME",
    GAME_FINISHED = "GAME_FINISHED",
    UPDATE_GAME = "UPDATE_GAME",
    UPDATE_PLAYERS = "UPDATE_PLAYERS",
    UPDATE_NOTEPADS = "UPDATE_NOTEPADS",
}

export interface setGameCode {
    type: ActionTypes.SET_GAME_CODE;
    gameCode: string;
    setNotepads: (notepads: Record<string, Notepad>) => updateNotepads;
    setPlayers: (players: Record<string, Player>) => updatePlayers;
    gameIsNew: boolean;
}

export interface viewPlayerHistory {
    type: ActionTypes.VIEW_PLAYER_HISTORY;
    playerId: UUID;
}

export interface viewNotepadHistory {
    type: ActionTypes.VIEW_NOTEPAD_HISTORY;
    ownerId: UUID;
}

export interface startGame {
    type: ActionTypes.START_GAME;
}

export interface gameFinished {
    type: ActionTypes.GAME_FINISHED;
}

export interface updateGame {
    type: ActionTypes.UPDATE_GAME;
    game: Game;
}

export interface updateNotepads {
    type: ActionTypes.UPDATE_NOTEPADS;
    notepads: Record<string, Notepad>;
}

export interface updatePlayers {
    type: ActionTypes.UPDATE_PLAYERS;
    players: Record<string, Player>;
}

export type Action =
    viewPlayerHistory |
    viewNotepadHistory |
    startGame |
    gameFinished |
    setGameCode |
    updateGame |
    updateNotepads |
    updatePlayers;

export interface Actions extends Record<string, (...args: any) => Action> {
    setGameCode: (
        gameCode: string,
        setNotepads: (notepads: Record<string, Notepad>) => updateNotepads,
        setPlayers: (players: Record<string, Player>) => updatePlayers,
        gameIsNew?: boolean
    ) => setGameCode;
    viewPlayerHistory: (playerId: UUID) => viewPlayerHistory,
    viewNotepadHistory: (ownerId: UUID) => viewNotepadHistory,
    startGame: () => startGame,
    updateGame: (game: Game) => updateGame,
    updateNotepads: (notepads: Record<string, Notepad>) => updateNotepads,
    updatePlayers: (players: Record<string, Player>) => updatePlayers,
    gameFinished: () => gameFinished,
}

export type Store = [State, Actions];
