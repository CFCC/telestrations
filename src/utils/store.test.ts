import * as store from "./store";

jest.mock("./firebase");

it("Default State", () => {
    expect(store.defaultState).toMatchSnapshot();
});

it("Epic", () => {

});

describe("Reducer", () => {

});
