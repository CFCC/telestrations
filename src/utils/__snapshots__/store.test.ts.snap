// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default State 1`] = `undefined`;
